drivers/counter/counter_nrfx_timer.c

```diff
@@ -55,112 +65,173 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
 				struct counter_nrfx_config, info);
 }
 
-static int counter_nrfx_start(struct device *dev)
+static int start(struct device *dev)
 {
-	nrfx_timer_enable(&get_nrfx_config(dev)->timer);
+	nrf_timer_task_trigger(get_nrfx_config(dev)->timer,
+			       NRF_TIMER_TASK_START);
 
 	return 0;
 }
 
-static int counter_nrfx_stop(struct device *dev)
+static int stop(struct device *dev)
 {
-	nrfx_timer_disable(&get_nrfx_config(dev)->timer);
+	nrf_timer_task_trigger(get_nrfx_config(dev)->timer,
+				NRF_TIMER_TASK_SHUTDOWN);
 
 	return 0;
 }
 
-static u32_t counter_nrfx_get_top_value(struct device *dev)
+static u32_t get_top_value(struct device *dev)
+{
+	return nrf_timer_cc_read(get_nrfx_config(dev)->timer, TOP_CH);
+}
+
+static u32_t get_max_relative_alarm(struct device *dev)
 {
-	return nrfx_timer_capture_get(&get_nrfx_config(dev)->timer, TOP_CH);
+	return get_top_value(dev);
 }
 
-static u32_t counter_nrfx_get_max_relative_alarm(struct device *dev)
+static u32_t read(struct device *dev)
 {
-	return nrfx_timer_capture_get(&get_nrfx_config(dev)->timer, TOP_CH);
+	NRF_TIMER_Type *timer = get_nrfx_config(dev)->timer;
+
+	nrf_timer_task_trigger(timer,
+			       nrf_timer_capture_task_get(COUNTER_READ_CC));
+
+	return nrf_timer_cc_read(timer, COUNTER_READ_CC);
 }
 
-static u32_t counter_nrfx_read(struct device *dev)
+/* Return true if value equals 2^n - 1 */
+static inline bool is_power_of_2(u32_t val)
 {
-	return nrfx_timer_capture(&get_nrfx_config(dev)->timer,
-				  COUNTER_READ_CC);
+	return !(val & (val + 1));
 }
 
-/** @brief Calculate compare value.
- *
- * If ticks are relative then compare value must take into consideration
- * counter wrapping.
- *
- * @return Compare value to be used in TIMER channel.
- */
-static inline u32_t counter_nrfx_get_cc_value(struct device *dev,
-				const struct counter_alarm_cfg *alarm_cfg)
+static u32_t ticks_add(u32_t val1, u32_t val2, u32_t top)
 {
-	u32_t remainder;
-	u32_t cc_val;
-	u32_t ticks = alarm_cfg->ticks;
+	u32_t sum = val1 + val2;
 
-	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) != 0) {
-		return ticks;
+	if (likely(is_power_of_2(top))) {
+		sum = sum & top;
+	} else {
+		sum = sum > top ? sum - top : sum;
 	}
 
-	cc_val = counter_nrfx_read(dev);
-	remainder = counter_nrfx_get_top_value(dev) - cc_val;
+	return sum;
+}
+
+static u32_t ticks_sub(u32_t val, u32_t old, u32_t top)
+{
+	if (likely(is_power_of_2(top))) {
+		return (val - old) & top;
+	}
+
+	/* if top is not 2^n-1 */
+	return (val > old) ? (val - old) : val + top + 1 - old;
```

Comment:
It should be `(val >= old)`, otherwise `top + 1` can be returned.

Reply:
fixed